dsa-23au/DSAFormat.md

```diff
@@ -0,0 +1,231 @@
+
+```
+<algo type>
+l: <line numbers>
+id: 
+n: <size of dataset>
+<body>
+...
+... <end-of-data after <l> lines of body>
+```
+
+Each of Data Structure and Algorithm categories contains a snapshot type and an update type.
+The first time an entity appears in a history, it is a snapshot type, to give a complete 
+description of the entity at time step 0. At every timestep later, only an update type frame 
+is needed to indicate what has changed since the last frame.
+
+## Algorithms 
+
+Algorithm entities consist of lines of pseudocode, variables, inputs, outputs, and a current 
+line of execution, which starts at 0.
+
+The lines of pseudocode may be of several types:
+* an assignment of a value (or a reference)
+* an assignment with a function call, which may be the current function (a recursive call) or another function 
+  * SWS functions are lazily evaluated, so you may refer to a function that is yet to be defined
+* an if statement with a boolean condition, opening a code block
+* a while / for loop, with a boolean condition, opening a code block
+* ending a code block
+* return values from the function
+
+```
+Function
+id: zzz 
+name: <name of algo> 
+args: <number of args>
+arg1: 
+arg2:
+l: <number of lines>
+<statement1>
+...
+<statementl>
+```
+
+### Step Frame
+
+This is SWS's main joy in life: to step through an algorithm. 
+These step frames tell the displayer how to update its display of both the
+algorithm (any functions currently defined and steppable through)
+and any data structures it is operating on.
+
+```
+Step
+count: <step_count>
+nextFunc: <id of function at end of this step>
+nextLine: <line number of function at end of this step>
+returnVal: <value to assign variable on current line>
+l: 3 
+<id1 of ds>: highlight <index1> # highlight the value at index1 
+<id2 of ds>: value <index2> <value2> # change the value at index2 to value2
+<id3 of ds>: highlight <index3> value <index4> <value4>
+```
+### Assignment Statement 
+
+Adds a new variable called `var1` to the function scope, initialized to `value1`.
+
+```
+var1 <-- value 
+```
+
+### Function Call 
+
+```
+var1, var2 <-- call <function id> arg1, arg2, ...
+```
+
+### If Statement 
+
+```
+if <expression1> 
+    <body>
+end
+```
+
+### While Loop 
+
+```
+while <expression>
+```
+
+### Do While Loop
+
+Begins a `while` loop which executes at least once, and evaluates its continue condition.
+
+```
+do 
+    <body>
+while <expression>
+```
+
+### End Code block
+
+Ends a code block begun by an `if` statement or a `while / for` loop.
+
+## Data Structures
+
+There are x built-in data structures in this first version, and more may be added
+in the future. Each can be oriented "Down", meaning drawn vertically with increasing indices
+going down, or "Right", meaning draw horizontally with increasing indices going to the right.
+
+```
+<ds type>
+id: zzz
+l: xx
+n: yyy
+<body>
+...
+... <end-of-data after <l> lines of body>
+```
+
+### Arrays
```

Comment:
for comparing data pulled from Spotify, would arrays be the best choice?

Reply:
Yes, most data starts off as being a list / array, especially when it's in CSV format.
From the Spotify, the data they send you will probably be in CSV, or JSON, if you pull it off the web.